test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -81,23 +90,25 @@ class SavedMatrix(matrix: TestMatrix) {
         outcome = success
         if (matrix.testExecutions == null) return
 
-        matrix.testExecutions.forEach {
-            val executionResult = GcToolResults.getExecutionResult(it)
-            updateOutcome(executionResult.outcome)
-
-            // testExecutionStep, testTiming, etc. can all be null.
-            // sometimes testExecutionStep is present and testTiming is null
-            val stepResult = GcToolResults.getStepResult(it.toolResultsStep)
-            val testTimeSeconds = stepResult.testExecutionStep?.testTiming?.testProcessDuration?.seconds ?: return
-
-            val billableMinutes = Billing.billableMinutes(testTimeSeconds)
+        updateFinishedMatrixData(matrix)
+    }
 
-            if (AndroidCatalog.isVirtualDevice(it.environment?.androidDevice, matrix.projectId ?: "")) {
-                billableVirtualMinutes += billableMinutes
-            } else {
-                billablePhysicalMinutes += billableMinutes
+    private fun updateFinishedMatrixData(matrix: TestMatrix) {
+        matrix.testExecutions.createTestExecutionDataListAsync()
+            .map {
+                FinishedTestMatrixData(
```

Comment:
Right it would be better to have function instead of data class:
```
private fun updateFinishedMatrixData(matrix: TestMatrix) =
        matrix.testExecutions.createTestExecutionDataListAsync().forEach {
            updateAllOutcomes(
                stepOutcome = GcToolResults.getExecutionResult(it.testExecution).outcome,
                isVirtualDevice = AndroidCatalog.isVirtualDevice(
                    it.testExecution.environment.androidDevice,
                    matrix.projectId.orEmpty()
                ),
                testSuiteOverviewData = it.createTestSuitOverviewData(),
                billableMinutes = it.step.testExecutionStep?.testTiming?.testProcessDuration?.seconds
                    ?.let { testTimeSeconds -> Billing.billableMinutes(testTimeSeconds) }
            )
        }

private fun updateAllOutcomes(
    stepOutcome: Outcome,
    isVirtualDevice: Boolean,
    testSuiteOverviewData: TestSuiteOverviewData?,
    billableMinutes: Long?
) {
    updateOutcome(stepOutcome)
    updateOutcomeDetails(stepOutcome, testSuiteOverviewData)
    billableMinutes?.let { updateBillableMinutes(it, isVirtualDevice) }
}
```

Reply:
I change the proposed name from `updateAllOutcomes ` to `updatedFinishedInfo`, cause we update single outcome, details and minutes not all outcomes